### ISSUE TITLE
SSH client only uses the deprecated MSG_KEX_DH_GEX_REQUEST_OLD instead of using MSG_KEX_DH_GEX_REQUEST

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8100]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)| @cjwatson reported|
|-|-|
|Trac ID|trac#8100|
|Type|enhancement|
|Created|2015-11-11 10:21:38Z|
|Branch|https://github.com/twisted/twisted/tree/ssh-client-kex-old-8100-2|

The diffie-hellman-group-exchange-sha256 support I added in [#7672](https://github.com/twisted/twisted/issues/7672) interoperates as a client with OpenSSH servers up to 6.8, but not 6.9 or newer.  This is because it uses MSG_KEX_DH_GEX_REQUEST_OLD to initiate group exchange.  This message number is listed as "for backward compatibility" in RFC 4419 (March 2006), probably because the newer MSG_KEX_DH_GEX_REQUEST is more useful: it allows specifying a range of acceptable group sizes rather than simply a preferred size.  OpenSSH dropped support for this in 6.9 with the commit message "deprecate ancient, pre-RFC4419 and undocumented SSH2_MSG_KEX_DH_GEX_REQUEST_OLD message".

The chances of running across a server implementation that supports DH group exchange but not MSG_KEX_DH_GEX_REQUEST are pretty slim.  Support for MSG_KEX_DH_GEX_REQUEST was added in OpenSSH 2.9p2 at the latest (probably earlier - git history that far back is confusing to navigate), and OpenSSH's compat.c records no non-OpenSSH server implementations that use the old message type.  I think that therefore it is entirely safe to switch to the new message type in the conch client.

Attachments:

* [drop-old-dhgex-request.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94e/94ea7db71daea81d8b156eff10e4e1621414ebfe/2a09a720881a07a5a9e74593dd12a5f9b46244ac.patch) (3482 bytes) - added by cjwatson on 2015-11-11 10:24:08Z - 
* [drop-old-dhgex-request-1-to-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94e/94ea7db71daea81d8b156eff10e4e1621414ebfe/74e7db92a1be4ad57383af6a7895e9c1d0dee594.patch) (3657 bytes) - added by cjwatson on 2015-11-12 11:31:53Z - Patch against 0406b559dd8432bdf355a30611f5f0d8056dee59 to make instance variables private and document them
* [drop-old-dhgex-request-2-to-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/94e/94ea7db71daea81d8b156eff10e4e1621414ebfe/6a21161d04e9eb9e96a8ced121302db1dea06e5e.patch) (757 bytes) - added by cjwatson on 2015-11-12 11:35:20Z - Expand comment

<details><summary>Searchable metadata</summary>

```
trac-id__8100 8100
type__enhancement enhancement
reporter__cjwatson cjwatson
priority__normal normal
milestone__None None
branch__branches_ssh_client_kex_old_8100_2 branches/ssh-client-kex-old-8100-2
branch_author__adiroiban adiroiban
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__None None
time__1447237298984296 1447237298984296
changetime__1447361085147634 1447361085147634
version__None None
owner__cjwatson cjwatson
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_2"></a>|@cjwatson commented|
|-|-|

This patch switches to the new message for initiating DH group exchange.  Server support for the old message is retained.  I've interop-tested this against OpenSSH 6.7 and 6.9 servers.

Minimum and maximum DH group sizes are from the RFC, and OpenSSH uses the same fixed values for these.  The preferred DH group size ought to be computed more dynamically as OpenSSH does, but this is no worse than the previous state; I've just made it clearer what's going on (by using struct.pack rather than a hardcoded byte string) and left an XXX comment.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban commented|
|-|-|

I agree that we should start with MSG_KEX_DH_GEX_REQUEST.

It looks like there is a bug in OpenSSH https://bugzilla.mindrot.org/show_bug.cgi?id=2494 ... so for OpenSSH we can not fallback or try other method

I don't know what to say about the backward compatibility.

I would say that this is a bug in conch.ssh.client. Based on RFC 4419 the client should use MSG_KEX_DH_GEX_REQUEST

I think that conchs.sshclient might use MSG_KEX_DH_GEX_REQUEST_OLD for backward compatiblity... in case SSH_MSG_UNIMPLEMENTED is returned from MSG_KEX_DH_GEX_REQUEST

again, the RFC 4419 is very brief here :(

I could not find any reference about pre-4419 support in client docs/docstring... so I don't know if this change qualifies for compatible or incompatible.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

(In [46139]) Branching to ssh-client-kex-old-8100.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

From IRC
```
(14:53:44) glyph: adiroiban: definitely compatible
(14:53:58) glyph: adiroiban: the most important compatibility to preserve is python-API level
(14:54:11) glyph: adiroiban: literally nobody in the world should be running openssh 2.9p2
```

we can do this without deprecation work.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Hi Colin and many thanks for your patch.

all new code should be documented, this include instance/class variables ... even if they are private.

I tried to updated your patch to document them.

We can send the recommended OpenSSH values, but if our client only support 1024 primes, then we should not ask for 2048 or inform the server that we are ok with bigger numbers.

So I think that this patch should include handling for the mix/max/pref values.

I tried to update your patch to sketch one possible fix.

There are 2 commits... first of them is exposing a new public API to allow configuring the preferred size.

I think that we start without any public API change and just document the rule based on which the preferred method is chose... later we can see how to expose this as a public API.

With the preferred size we hit the configuration area... and this is delicate issue in Twisted :(

Please review my comments and commits.

Either submit a new patch or come back with feedback.

Thanks again and sorry for all the noise related to the compatibility policy.

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_7"></a>|@cjwatson commented|
|-|-|

A couple of typos in your recent changes:

```
+            # Maybe no commons protocols were agreed.
```

commons -> common

```
+            # We agreed on a fixed group key exchange algoright.
```

algoright -> algorithm

Your changes to document instance variables are fine, and of course it's better for them to be private.

Regarding configuration, I think you're on the wrong path.  This doesn't actually need to be configurable: that would be overkill and the caller doesn't have easy access to the information needed to work out a sensible value anyway.  What I meant by my comment that it should be computed more dynamically is that Twisted should work out a safer preferred group size **for itself**.  The way that OpenSSH does this - and I expect this is from a cryptographic standard/paper somewhere, but I don't have the reference to that - is to take the maximum of the encryption cipher key length, the encryption block size, the encryption IV length, and the MAC key length, which gives it a number of bits; and then it estimates a DH group size with an attack complexity approximately O(2^bits^), which is a simple mapping that you can find in its `dh_estimate` function.

So, I think we should **not** expose a new public API to make this configurable, but rather we should have a new ticket for conch to work out a better preferred group size for itself, and then leave it at that.  Don't make things configurable where Twisted can do a better job of working out good values for itself.

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)|@cjwatson set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_9"></a>|@cjwatson commented|
|-|-|

The two attachments apply in sequence on top of my initial patch to fix up instance variable handling, per adiroiban's review.  I've filed [#8103](https://github.com/twisted/twisted/issues/8103) for the preferred group size issue.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_10"></a>|@adiroiban commented|
|-|-|

(In [46158]) Branching to ssh-client-kex-old-8100-2.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_11"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Thanks. Changes look good.

I added the missing news file fragments to advertise the changes.

I don't think that this is a bug... but rather a feature which allows Conch SSH client to work with latest OpenSSH versions. Please correct me if I am wrong :)

I added the standard comment format for FIXME reference to [#8103](https://github.com/twisted/twisted/issues/8103).

I moved var initialization in the class definition. If someone wants to go with different values they can overwrite those values... not recommended... also because those are private.

I have sent the changes to buildbot.

Please check my latest commits and make sure they make sense.

If all is OK, I will merge it.

Many, many thanks for your work on this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="cjwatson's avatar" src="https://avatars.githubusercontent.com/u/2429141?s=50" width="50" height="50">](https://github.com/cjwatson)<a name="note_12"></a>|@cjwatson set owner to @adiroiban|
|-|-|

Your changes look fine, thank you, except that you have written "support" in twisted/conch/topfiles/8100.removal where it should be "supports" (the class is singular).  I would suggest that 8100.feature should probably add "using MSG_KEX_DH_GEX_REQUEST" after "key exchange" too; your sentence is strictly correct if you know the intricacies of RFC 4419, but adding the message name makes it clearer and parallels 8100.removal.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_13"></a>|@adiroiban set owner to @cjwatson|
|-|-|

Many thanks for your comments.

Waiting for the builds and then will merge.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_14"></a>|@adiroiban set status to `closed`|
|-|-|

(In [46168]) Merge ssh-client-kex-old-8100-2: Add support in twisted.conch.ssh client for RFC 4419.

Author: cjwatson
Reviewer: adiroiban
Fixes: [#8100](https://github.com/twisted/twisted/issues/8100)
